pkg/gather/azure/azure.go

```diff
@@ -0,0 +1,293 @@
+		authorizer:        session.Authorizer,
+		environment:       session.Environment,
+		resourceGroupName: resourceGroupName,
+		logger:            logger,
+		serialLogBundle:   serialLogBundle,
+		bootstrap:         bootstrap,
+		masters:           masters,
+		directory:         filepath.Dir(serialLogBundle),
+		cloudName:         cloudName,
+	}
+
+	gather.virtualMachinesClient = compute.NewVirtualMachinesClientWithBaseURI(session.Environment.ResourceManagerEndpoint, gather.subscriptionID)
+	gather.virtualMachinesClient.Authorizer = gather.authorizer
+
+	gather.accountsClient = storage.NewAccountsClientWithBaseURI(session.Environment.ResourceManagerEndpoint, gather.subscriptionID)
+	gather.accountsClient.Authorizer = gather.authorizer
+
+	return gather, nil
+}
+
+// Run is the entrypoint to start the gather process.
+func (g *Gather) Run() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	credentials, err := getSharedKeyCredentials(ctx, g)
+	if err != nil {
+		return err
+	}
+
+	// XXX: we can probably filter here on Hostname/IP address
+	virtualMachines, err := getVirtualMachines(ctx, g)
+	if err != nil {
+		return err
+	}
+
+	var files []string
+
+	// We can only get the serial log from VM's with boot diagnostics enabled
+	bootDiagnostics := getBootDiagnostics(ctx, virtualMachines, g)
+	for _, bootDiagnostic := range bootDiagnostics {
+		screenBmp := to.String(bootDiagnostic.ConsoleScreenshotBlobURI)
+		files = append(files, screenBmp)
+
+		serialLog := to.String(bootDiagnostic.SerialConsoleLogBlobURI)
+		files = append(files, serialLog)
+	}
+
+	err = downloadFiles(ctx, files, credentials, g)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getSharedKeyCredentials(ctx context.Context, g *Gather) ([]*azblob.SharedKeyCredential, error) {
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+
+	accountListResult, err := g.accountsClient.ListByResourceGroup(ctx, g.resourceGroupName)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not find any storage accounts")
+	}
+
+	var credentials []*azblob.SharedKeyCredential
+	if accountListResult.Value != nil {
+		for _, account := range *accountListResult.Value {
+			accountName := to.String(account.Name)
+			keyResults, err := g.accountsClient.ListKeys(ctx, g.resourceGroupName, accountName, "")
+			if err != nil {
+				g.logger.Debugf("failed to list keys: %s", err.Error())
+				continue
+			}
+			if keyResults.Keys != nil {
+				for _, key := range *keyResults.Keys {
+					if key.Value != nil {
+						credential, err := azblob.NewSharedKeyCredential(accountName, to.String(key.Value))
+						if err != nil {
+							g.logger.Debugf("failed to get shared key: %s", err.Error())
+							continue
+						}
+						credentials = append(credentials, credential)
+					}
+				}
+			}
+		}
+	}
+
+	return credentials, nil
+}
+
+func getVirtualMachines(ctx context.Context, g *Gather) ([]compute.VirtualMachine, error) {
```

Comment:
The approach of fetching all machines works for Azure, where everything is contained in a resource group specific to the cluster. However, for other platforms that do not have the concept of a resource group, we will need filtering on the virtual machines. For Azure, we would grab boot diagnostics for workers. For other platforms, we would presumably not grab boot diagnostics for workers. I would rather see a consistent approach across platforms.